third_party/WebKit/Source/core/loader/FrameLoader.cpp

```diff
@@ -1218,8 +1218,12 @@ String FrameLoader::userAgent() const
     String clientUserAgent = client()->userAgent();
 
     // On Endless, we need to still be able to alter the UserAgent string depending
-    // on the URL we are loading, so we adapt it if needed before applying it.
-    String userAgent = document ? adaptUserAgentForURL(clientUserAgent, document->url()) : clientUserAgent;
+    // on the URL we are loading, so we adapt it if needed before applying it. But we
+    // might not have a committed load, so we first check the URL from the provisional
+    // loader before trying with the regular one, so that we always get a valid URL.
+    const DocumentLoader& activeLoader = m_provisionalDocumentLoader.get() ? *m_provisionalDocumentLoader : *m_documentLoader;
+    const KURL& url = activeLoader.request().url();
+    String userAgent = document ? adaptUserAgentForURL(clientUserAgent, url) : clientUserAgent;
```

Comment:
> If this were an upstream review I would say this code belongs in the FrameLoaderClient, but I think this is totally fine for Endless.

I thought the same at the beginning, but after some experiments back in the day when I wrote this I found that doing it that way would work, because of how the separation between chrome and blink is. But you might be right and I might have overlooked something, although I also agree with you that this is totally fine for Endless, so I think/hope it's possibly good enough as it is, since it works and is clean as a downstream patch.

> I did not check to ensure m_documentLoader is guaranteed to be valid if m_provisionalDocumentLoader is NULL; I hope that is true?

The non-provisional loader is set by stealing the reference from the provisional one by calling `m_provisionalDocumentLoader.release()` in `FrameLoader::prepareForCommit()`, so it is a true assumption indeed.